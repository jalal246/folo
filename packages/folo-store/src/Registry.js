class Registry {
  constructor() {
    this.btnGroup = new Set();
    this.dataObj = {};
    this.triggers = {};
  }

  getStorRef(storeID = "unrecognized", isCheck = false) {
    const $ = this.dataObj;

    if (isCheck && !$[storeID]) {
      $[storeID] = {};
    }

    return $[storeID];
  }

  /**
   * Add cell to dataObj
   * Add groupName to btnGroup
   *
   * This function will be called when cells mount
   * after componentDidMount, the data in dataObj will be moved to state
   * dataObj is a temp object holds value to avoid update state while rendering
   * in this case will init all cells in dataObj until rendering happens
   * then update the state so all the values update happen in state
   *
   * @param {object} cell - new cell that should be register
   * @param {string} cell.nameRef   key for value
   * @param {string||boolean} cell.initValue value
   * @param {string} cell.groupName group name in case the cell is group-toggle
   */
  subscribe({ nameRef, initValue, groupName, storeID }, updater) {
    const storeRef = this.getStorRef(storeID, true);

    storeRef[nameRef] = initValue;

    // if it has group, handle it
    if (groupName) {
      this.triggers[nameRef] = updater;

      /*
       * check if group name not exist then add it and create its own set
       * then add the cell to its group
       */

      if (!this.btnGroup.has(groupName)) {
        // add it because it is new group
        this.btnGroup.add(groupName);

        // create new set for the group
        this.btnGroup[groupName] = new Set();
      }

      if (!this.btnGroup[groupName].has(nameRef)) {
        // then add the cell name to where its belong
        // to its group
        this.btnGroup[groupName].add(nameRef);
      }
    }
  }

  updater({ nameRef, newValue, groupName, storeID }) {
    const storeRef = this.getStorRef(storeID);

    storeRef[nameRef] = newValue;

    if (groupName) {
      if (newValue !== false) {
        // update group of values

        // toggle group values
        this.btnGroup[groupName].forEach((FieldNameRef) => {
          // toggle all except the targeted key name which called nameRef
          // since we already changed its value above
          if (FieldNameRef !== nameRef) {
            storeRef[FieldNameRef] = !newValue;
            this.triggers[FieldNameRef](!newValue);
          }
        });
      }
    }
  }

  getAll(storeID) {
    const storeRef = this.getStorRef(storeID);

    return storeRef;
  }

  clear(storeID) {
    if (!storeID) {
      this.dataObj.unrecognized = {};
    } else {
      this.dataObj[storeID] = {};
    }
  }
}

export default Registry;
