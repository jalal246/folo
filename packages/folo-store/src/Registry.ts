type StoreValue = string | Boolean;

interface SubscribeInfo {
  nameRef: string;
  initValue: StoreValue;
  groupName?: string;
  storeID?: string;
}

interface Updater {
  nameRef: string;
  newValue: StoreValue;
  groupName?: string;
  storeID?: string;
}

interface DataHolder {
  [key: string]: { [key: string]: Object };
}

interface Triggers {
  [key: string]: { [key: string]: Function };
}

interface BtnGroup {
  [key: string]: { [key: string]: Set<string> };
}

const DEFAULT_STORE = "unrecognized";

class Registry {
  dataHolder: DataHolder;

  triggers: Triggers;

  activeStore: string;

  activeField: string;

  btnGroup: BtnGroup;

  constructor() {
    this.dataHolder = {};
    this.triggers = {};

    this.btnGroup = {};
    this.activeStore = "";
    this.activeField = "";
  }

  setActive(nameRef?: string, storeID?: string) {
    this.activeStore = storeID || DEFAULT_STORE;
    if (nameRef) this.activeField = nameRef;
  }

  assignValueToStore(storeValue: StoreValue) {
    if (!this.dataHolder[this.activeStore]) {
      this.dataHolder[this.activeStore] = {};
      this.btnGroup[this.activeStore] = {};
    }

    this.dataHolder[this.activeStore][this.activeField] = storeValue;
  }

  assignTrigger(triggerHandler: Function) {
    this.triggers[this.activeStore][this.activeField] = triggerHandler;
  }

  triggerHandler(withValue: StoreValue) {
    this.triggers[this.activeStore][this.activeField](withValue);
  }

  /**
   * Add field to dataBranches
   * Add groupName to btnGroup
   *
   * This function will be called when cells mount
   * after componentDidMount, the data in dataBranches will be moved to state
   * dataBranches is a temp object holds value to avoid update state while rendering
   * in this case will init all cells in dataBranches until rendering happens
   * then update the state so all the values update happen in state
   *
   * @param {object} field - new field that should be register
   * @param {string} field.nameRef   key for value
   * @param {string||boolean} field.initValue value
   * @param {string} field.groupName group name in case the field is group-toggle
   */
  subscribe(
    { nameRef, initValue, groupName, storeID }: SubscribeInfo,
    triggerHandler: Function
  ) {
    this.setActive(nameRef, storeID);

    this.assignValueToStore(initValue);

    // if it has group, handle it
    if (groupName) {
      this.assignTrigger(triggerHandler);

      if (!this.btnGroup[this.activeStore][groupName]) {
        /*
         * Check if group name not exist then add it and create its own set
         * then add the field to its group
         */

        // create new set for the group
        this.btnGroup[this.activeStore][groupName] = new Set();
      }

      if (!this.btnGroup[this.activeStore][groupName].has(nameRef)) {
        // then add the field name to where its belong
        // to its group
        this.btnGroup[this.activeStore][groupName].add(nameRef);
      }
    }
  }

  updater({ nameRef, newValue, groupName, storeID }: Updater) {
    this.setActive(nameRef, storeID);
    this.assignValueToStore(newValue);

    if (groupName) {
      if (newValue !== false) {
        // update group of values

        // toggle group values
        this.btnGroup[this.activeStore][groupName].forEach((FieldNameRef) => {
          // toggle all except the targeted key name which called nameRef
          // since we already changed its value above
          if (FieldNameRef !== nameRef) {
            this.setActive(FieldNameRef, storeID);
            this.assignValueToStore(!newValue);
            this.triggerHandler(!newValue);
          }
        });
      }
    }
  }

  getDataByStoreID(storeID?: string): Object {
    this.setActive(undefined, storeID);

    return this.dataHolder[this.activeStore];
  }

  clear(storeID?: string) {
    this.setActive(undefined, storeID);

    this.dataHolder[this.activeStore] = {};
    this.triggers[this.activeStore] = {};
    this.btnGroup[this.activeStore] = {};
  }

  destroy() {
    this.dataHolder = {};
    this.triggers = {};
    this.btnGroup = {};
  }
}

export default Registry;
