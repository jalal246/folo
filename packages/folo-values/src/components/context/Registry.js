class Registry {
  constructor() {
    this.btnGroup = new Set();
    this.datatObj = {};

    this.registerCellInfo = this.registerCellInfo.bind(this);
  }

  /**
   * Add cell to datatObj
   * Add groupName to btnGroup
   *
   * This function will be called when cells mount
   * after componentDidMount, the data in datatObj will be moved to state
   * datatObj is a temp object holds value to avoid update state while rendeing
   * in this case will init all cells in datatObj until rendering happens
   * then update the state so all the values update happen in state
   *
   * @param {object} cell - new cell that should be register
   * @param {string} cell.nameRef   key for value
   * @param {string||boolean} cell.initValue value
   * @param {string} cell.groupName group name in case the cell is group-toggle
   */
  registerCellInfo({ nameRef, initValue, groupName }) {
    this.datatObj[nameRef] = initValue;

    // if it has group, handle it
    if (groupName) {
      /*
       * check if group name not exist then add it and create its own set
       * then add the cell to its group
       */

      if (!this.btnGroup.has(groupName)) {
        // add it because it is new groupÿ°
        this.btnGroup.add(groupName);

        // create new set for the group
        this.btnGroup[groupName] = new Set();
      }

      if (!this.btnGroup[groupName].has(nameRef)) {
        // then add the cell name to where its belong
        // to its group
        this.btnGroup[groupName].add(nameRef);
      }
    }
  }
}

export default Registry;
