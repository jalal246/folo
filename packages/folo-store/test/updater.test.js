import Registry from "../src";

let registry;

const NAME_REF_1 = "input1";

const NAME_REF_2 = "input2";
const NAME_REF_3 = "input3";

const BTN_GROUP = "group1";

const btnGroup = new Set();
btnGroup.add(BTN_GROUP);
btnGroup.add(BTN_GROUP);
btnGroup.group1 = new Set();
btnGroup[BTN_GROUP].add(NAME_REF_2);
btnGroup[BTN_GROUP].add(NAME_REF_3);

const GENERAL_STORE_ID = "unrecognized";

const TEXT_NEW_VALUE = "some value";

const updater1 = jest.fn((newValue) => newValue);
const updater2 = jest.fn((newValue) => newValue);

describe("Registry updater", () => {
  beforeAll(() => {
    registry = new Registry();

    const input1 = {
      nameRef: NAME_REF_1,
      initValue: "",
      groupName: null,
    };
    registry.subscribe(input1);

    const input2 = {
      nameRef: NAME_REF_2,
      initValue: false,
      groupName: BTN_GROUP,
    };
    registry.subscribe(input2, updater1);

    const input3 = {
      nameRef: NAME_REF_3,
      initValue: false,
      groupName: BTN_GROUP,
    };
    registry.subscribe(input3, updater2);
  });

  it("Updates string value", () => {
    registry.updater({
      nameRef: NAME_REF_1,
      newValue: TEXT_NEW_VALUE,
    });

    expect(registry.dataObj).toStrictEqual({
      [GENERAL_STORE_ID]: {
        [NAME_REF_1]: TEXT_NEW_VALUE,
        [NAME_REF_2]: false,
        [NAME_REF_3]: false,
      },
    });

    expect(registry.dataObj).toMatchSnapshot();
  });

  it("Toggles boolean value in a group button", () => {
    registry.updater({
      nameRef: NAME_REF_2,
      newValue: true,
      groupName: BTN_GROUP,
    });

    expect(registry.dataObj).toStrictEqual({
      [GENERAL_STORE_ID]: {
        [NAME_REF_1]: TEXT_NEW_VALUE,
        [NAME_REF_2]: true,
        [NAME_REF_3]: false,
      },
    });

    expect(updater1).toHaveBeenCalledTimes(0);
    expect(updater2).toHaveBeenCalledWith(false);

    expect(registry.dataObj).toMatchSnapshot();
  });
});
