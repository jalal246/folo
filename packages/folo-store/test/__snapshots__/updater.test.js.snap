// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry updater Toggles boolean value in a group button 1`] = `
Object {
  "unrecognized": Object {
    "input1": "some value",
    "input2": true,
    "input3": false,
  },
}
`;

exports[`Registry updater Updates string value 1`] = `
Object {
  "unrecognized": Object {
    "input1": "some value",
    "input2": false,
    "input3": false,
  },
}
`;
